Engine traits cleanup

brson opened the pull request:
Polish up the engine_traits and engine_rocks crates to use consistent organization, naming, and patterns. Add docs.

###  What have you changed?

This is mostly taking what I've learned and implemented in https://github.com/tikv/tikv/pull/5657 and doing the same thing consistently across the engine_traits and engine_rocks traits, and documenting what I'm doing so others can do the same.

This extends https://github.com/tikv/tikv/pull/5657, so most of the commits are from that PR. New commits start at "move db.rs to engine.rs".

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  How is the PR tested?

typechecking

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?


###  Does this PR affect `tidb-ansible`?

no

###  Refer to a related PR or issue link (optional)

https://github.com/tikv/tikv/issues/4184

###  Benchmark result if necessary (optional)

###  Any examples? (optional)

brson commented:
I'll rebase this tomorrow.

brson commented:
OK, this is rebased against master and now only includes new commits.

brson commented:
ptal @5kbpers @aknuds1 @overvenus

kennytm commented on components/engine_rocks/src/lib.rs:
```diff
@@ -1,170 +1,44 @@
+//! This is a work-in-progress attempt to abstract all the features needed by
+//! TiKV to persist its data.
+//!
+//! The module structure here mirrrors that in engine_traits where possible.
```
```suggestion
//! The module structure here mirrors that in engine_traits where possible.
```

kennytm commented on components/engine_traits/src/lib.rs:
```diff
@@ -1,79 +1,108 @@
+// they are required by KvEngine or are an associated type of KvEngine. It is
+// recommended that engines follow the same module layout.
+//
+// Many of these define "extension" traits, that end in `Ex`.
```
```suggestion
// Many of these define "extension" traits, that end in `Ext`.
```

Connor1996 commented on components/engine_rocks/src/engine.rs:
```diff
@@ -108,47 +80,31 @@ impl KvEngine for Rocks {
         self.0.cf_names()
     }
-
-    fn delete_all_in_range_cf(
```
why removing this?

brson replied:
From the commit message:

> The code is unused, is duplicated between engine_rocks and engine and is not the
    same. I'm going to come back to this when it actually needs to be used.

I'm suspicious of it.

zhangjinpeng1987 replied:
https://github.com/tikv/tikv/blob/master/src/raftstore/store/fsm/apply.rs#L1293 and https://github.com/tikv/tikv/blob/master/src/server/gc_worker.rs#L391 use delete_all_in_range_cf.

Connor1996 commented on components/engine_rocks/src/tests.rs:
```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
```
Is it necessary to move the tests to a separate file? seems it's not a TiKV convention.

brson replied:
I've moved them to engine.rs.

zhangjinpeng1987 replied:
I think it is ok.

5kbpers commented on components/engine_traits/src/import.rs:
```diff
@@ -34,3 +28,9 @@ pub trait Import: CFHandleExt {
     ) -> Result<()>;
 }
+
+pub trait IngestExternalFileOptions {
```
Shall we convert other options to traits to keep consistency in this PR?

zhangjinpeng1987 replied:
Maybe step by step is better.

zhangjinpeng1987 commented:
Please address comments.

zhangjinpeng1987 commented on components/engine_traits/src/lib.rs:
```diff
@@ -1,79 +1,108 @@
-}
+// In our tests, we found that if the batch size is too large, running delete_all_in_range will
+// reduce OLTP QPS by 30% ~ 60%. We found that 32K is a proper choice.
+pub const MAX_DELETE_BATCH_SIZE: usize = 32 * 1024;
```
I think this is a RocksDB relative variable, better to move to engine_rocks.

brson replied:
The const is duplicated from the engine crate and appears to be unused so I've just deleted it.

brson commented:
Comments addressed.

Connor1996 commented:
`delete_all_in_range_cf` is still removed from engine trait?

brson commented:
@Connor1996 I explained this a few places: I removed it because it was duplicating a function from the engine crate, while having a slightly different implementation, and no callers. It was untested code, so I removed it in favor of the version that was tested. When callers are converted to use the trait, then the implementation can be moved from the engine crate to the engine_rocks crate, leaving no code duplication, and ensuring the function stays correct.

Connor1996 commented:
@brson Oh sorry, didn't notice that. I was thought we should copy the implementation from the `engine` crate to the `engine_rocks` crate, so we can keep the `delete_all_in_range_cf` in the trait. And once the callers are converted to use the trait, we remove the `engine` crate totally.
What I want to express is we need it in the trait, so remove it now and then add it back is unnecessary. Anyway, it's fine for me after your explanation. 😉Good job!
